### PR TITLE
Add test to measure end-to-end performance on pinned / non-pinned memory

### PR DESCRIPTION
**Pull request by GuanLuo**
This PR is waiting on #700 to be able to run in CI, the following is the log generated from local run. The log only shows end-to-end latency, but still, it suggests that the performance gain from using pinned memory is not significant.
Each element is 4 bytes
```
libtorch ensemble 16384 elements

non-pinned


Sync Serial Run (grpc)
Total time: 466.906 ms
Mean time: 4.669 ms
Stddev: 0.229 ms

Async Serial Run (grpc)
Total time: 459.508 ms
Mean time: 4.595 ms
Stddev: 0.139 ms
pinned


Sync Serial Run (grpc)
Total time: 461.882 ms
Mean time: 4.618 ms
Stddev: 0.102 ms

Async Serial Run (grpc)
Total time: 461.178 ms
Mean time: 4.611 ms
Stddev: 0.113 ms
libtorch ensemble 1048576 elements

non-pinned


Sync Serial Run (grpc)
Total time: 1262.29 ms
Mean time: 12.622 ms
Stddev: 0.84 ms

Async Serial Run (grpc)
Total time: 1556.33 ms
Mean time: 15.563 ms
Stddev: 3.224 ms
pinned


Sync Serial Run (grpc)
Total time: 1583.2 ms
Mean time: 15.831 ms
Stddev: 1.193 ms

Async Serial Run (grpc)
Total time: 1938.45 ms
Mean time: 19.384 ms
Stddev: 3.498 ms
libtorch ensemble 2097152 elements

non-pinned


Sync Serial Run (grpc)
Total time: 2213.8 ms
Mean time: 22.137 ms
Stddev: 0.503 ms

Async Serial Run (grpc)
Total time: 2494.92 ms
Mean time: 24.949 ms
Stddev: 2.617 ms
pinned


Sync Serial Run (grpc)
Total time: 2717.24 ms
Mean time: 27.172 ms
Stddev: 0.858 ms

Async Serial Run (grpc)
Total time: 3118.11 ms
Mean time: 31.181 ms
Stddev: 4.277 ms
libtorch ensemble 4194304 elements

non-pinned


Sync Serial Run (grpc)
Total time: 4024.22 ms
Mean time: 40.242 ms
Stddev: 0.491 ms

Async Serial Run (grpc)
Total time: 4504.22 ms
Mean time: 45.042 ms
Stddev: 3.703 ms
pinned


Sync Serial Run (grpc)
Total time: 4982.97 ms
Mean time: 49.829 ms
Stddev: 1.677 ms

Async Serial Run (grpc)
Total time: 5457.07 ms
Mean time: 54.57 ms
Stddev: 4.141 ms
libtorch ensemble 8388608 elements

non-pinned


Sync Serial Run (grpc)
Total time: 8968.74 ms
Mean time: 89.687 ms
Stddev: 2.807 ms

Async Serial Run (grpc)
Total time: 12849.5 ms
Mean time: 128.494 ms
Stddev: 6.836 ms
pinned


Sync Serial Run (grpc)
Total time: 8977.8 ms
Mean time: 89.777 ms
Stddev: 2.886 ms

Async Serial Run (grpc)
Total time: 11682.5 ms
Mean time: 116.825 ms
Stddev: 10.022 ms
libtorch ensemble 16777216 elements

non-pinned


Sync Serial Run (grpc)
Total time: 15313.8 ms
Mean time: 153.137 ms
Stddev: 5.671 ms

Async Serial Run (grpc)
Total time: 21661 ms
Mean time: 216.609 ms
Stddev: 10.442 ms
pinned


Sync Serial Run (grpc)
Total time: 15629 ms
Mean time: 156.29 ms
Stddev: 6.203 ms

Async Serial Run (grpc)
Total time: 21908 ms
Mean time: 219.08 ms
Stddev: 11.167 ms
```

### CODE REVIEW

File: qa/L0_pinned_memory/test.sh [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    run_server
+    if (( $SERVER_PID == 0 )); then
+        echo -e "\n***\n*** Failed to start $SERVER\n***"
+        cat $SERVER_LOG
+        exit 1
+    fi
+
+    # 16k 1m 2m 4m 8m 16m elements
+    set +e
+    for TENSOR_SIZE in 16384 1048576 2097152 4194304 8388608 16777216; do
```
Comment (deadeyegoodwin) [inline qa/L0_pinned_memory/test.sh]: So what size do you find it to be beneficial?

Comment (GuanLuo) [reply qa/L0_pinned_memory/test.sh]: Starting from size larger than 8m elements (32 MB), the latency using pinned memory is smaller.

Comment (deadeyegoodwin) [reply qa/L0_pinned_memory/test.sh]: Did you see this paper: https://www.cs.virginia.edu/~mwb7w/cuda_support/pinned_tradeoff.html
Your result seems to align with it. What we need to do is preallocate the entire pool of pinned memory and the allocate out chunks. So the pinned manager needs to get smarted. Can you file a 19.11 ticket for that. As for 19.10 you should just set the threshold for when pinned memory is used to be only for >32MB sizes.

Comment (GuanLuo) [reply qa/L0_pinned_memory/test.sh]: Yes, the observation of this paper makes sense to me. Should I add the threshold to this PR or a separate one.

Comment (deadeyegoodwin) [reply qa/L0_pinned_memory/test.sh]: Separate one.

**Pull request merged**